### PR TITLE
#178 add a consentStringUpdated handler to write a .digitru.st cookie

### PR DESCRIPTION
**Pull request by samtingleff**
On the consentStringUpdated event and when consent is present for ID 64, create an invisible iframe to https://cdn.digitru.st/prod/1.5.21/cmp-iframe.html. This iframe makes an ajax request to https://cdn-cf.digitru.st/id/v1, which writes an identity cookie to the digitru.st domain.

### CODE REVIEW

File: src/lib/init.js [hunk 1]
```diff
@@ -49,10 +49,25 @@ export function init(configUpdates) {
...
 				const { addEventListener, getVendorConsents } = cmp.commands;
+				if (config.digitrust.pixel) {
+					addEventListener('consentStringUpdated', digitrustPixel);
+				}
 				if (config.digitrust.redirects) {
 					addEventListener('consentStringUpdated', digitrustRedirect);
 				}
 
+				function digitrustPixel() {
+					getVendorConsents([64], (result) => {
```
Comment (shackijj) [inline src/lib/init.js]: How about moving 64 to a constant? It's used several times in this file.

File: src/lib/init.js [hunk 2]
```diff
@@ -49,10 +49,25 @@ export function init(configUpdates) {
...
+						if (result && result.vendorConsents && result.vendorConsents[64]) {
+							let iframe = document.createElement('iframe');
```
Comment (shackijj) [inline src/lib/init.js]: Let's move the creation of an iframe to a separate utility function.

File: src/lib/init.js (hunk 1 above)
Comment (shackijj) [inline src/lib/init.js]: The same getVendorConsent call is used in the digitrustRedirect function. Why wouldn't we create a function which  calls the getVendorConsents and checks this condition:

`result && result.vendorConsents && result.vendorConsents[64]`

File: src/lib/init.js [hunk 3]
```diff
@@ -49,10 +49,25 @@ export function init(configUpdates) {
 
 				// set cookies on digitrust domain after consent submitted
 				const { addEventListener, getVendorConsents } = cmp.commands;
+				if (config.digitrust.pixel) {
+					addEventListener('consentStringUpdated', digitrustPixel);
```
Comment (shackijj) [inline src/lib/init.js]: The string "consentStringUpdated" might be a constant.

**Comment by shackijj** [comment]
This approach will not work in Safari >= 11 since third-party cookies are disabled by default.

Why do we need to send an ajax request in the frame? It seems that we can avoid sending the request by setting cookies on the frame's request.

**Comment by samtingleff** [comment]
I know this will not work in Safari >= 11. There is an option to rewrite links through a .digitru.st service which is our only approach for Safari.

> Why do we need to send an ajax request in the frame?

You could also ask why we need an iframe at all... it looks like Chrome blocks third party cookies on XHR requests even when `withCredentials` is set.

File: src/lib/init.js (hunk 3 above)
Comment (samtingleff) [reply src/lib/init.js]: I couldn't find it defined anywhere... not to say that it shouldn't be.

File: src/lib/init.js (hunk 1 above)
Comment (samtingleff) [reply src/lib/init.js]: moved into metadata.json

File: src/lib/init.js (hunk 2 above)
Comment (samtingleff) [reply src/lib/init.js]: moved into utils.js

File: src/lib/init.js (hunk 1 above)
Comment (samtingleff) [reply src/lib/init.js]: created `getDigiTrustVendorConsents`

**Comment by shackijj** [comment]
> It looks like Chrome blocks third party cookies on XHR requests even when `withCredentials` is set.

I don't think that this is true. I've double checked it. Here is the [link](https://github.com/shackijj/cookie-demo) to the repo where you can find the code which I used for setting cookies.

File: src/lib/init.js [hunk 4]
```diff
@@ -59,10 +59,7 @@ export function init(configUpdates) {
 				function digitrustPixel() {
 					getVendorConsents([metadata.digitrustVendorId], (result) => {
 						if (result && result.vendorConsents && result.vendorConsents[metadata.digitrustVendorId]) {
-							let iframe = document.createElement('iframe');
-							iframe.src = config.digitrustIframeUrl;
-							iframe.style.width = "0px";
-							iframe.style.height = "0px";
+							let iframe = createIframe(config.digitrustIframeUrl, "0px", "0px");
 							document.body.appendChild(iframe);
```
Comment (shackijj) [inline src/lib/init.js]: How about moving this call in the function? Also, I suggest chaning the function's name to 'createInvisibleIframe`.

File: src/lib/init.js [hunk 5]
```diff
@@ -56,21 +56,27 @@ export function init(configUpdates) {
...
+							failure();
 						}
 					});
 				}
 
+				function digitrustPixel() {
+					getDigiTrustVendorConsents( (success) => {
+						let iframe = createIframe(config.digitrustIframeUrl, "0px", "0px");
+						document.body.appendChild(iframe);
+					}, (fail) => { });
```
Comment (shackijj) [inline src/lib/init.js]: `(fail) => { }` is used two times. We can make parameters optional.

File: src/lib/init.js [hunk 6]
```diff
@@ -49,18 +49,36 @@ export function init(configUpdates) {
...
+				}
 				if (config.digitrust.redirects) {
 					addEventListener('consentStringUpdated', digitrustRedirect);
 				}
 
-				function digitrustRedirect() {
-					getVendorConsents([64], (result) => {
-						if (result && result.vendorConsents && result.vendorConsents[64]) {
-							window.location.replace(`${config.digitrustRedirectUrl}${encodeURIComponent(window.location.href)}`);
+				function getDigiTrustVendorConsents(success, failure) {
```
Comment (shackijj) [inline src/lib/init.js]: I think that this function will be more readable and mainainable if it's written in the following way:

```
const ifVendorIdIsSet = (id, getVendorConsents, callback) => () => {
    getVendorConsents(id, (result) => {
        get(result, `vendorConsent.${id}`) && isFunction(callback) && callback()
    })
}

const sendInvisiblePixel = () => {...}

const onConsentStringUpdated = ifVendorIdIsSet(
    metadata.digitrustVendorId,
    getVendorConsents
    sendInvisiblePixel)

addEventListener('consentStringUpdated', onConsentStringUpdated)
```

File: src/lib/init.js [hunk 7]
```diff
@@ -49,18 +49,36 @@ export function init(configUpdates) {
...
+						if (result && result.vendorConsents && result.vendorConsents[metadata.digitrustVendorId]) {
+							success();
+						} else {
+							failure();
 						}
 					});
 				}
 
+				function digitrustPixel() {
+					getDigiTrustVendorConsents( (success) => {
```
Comment (shackijj) [inline src/lib/init.js]: the success argument is not used

**Pull request closed without merging**